### PR TITLE
LLVM port from release/13.x to release_13x

### PR DESCRIPTION
**Pull request by abrahamtovarmob**
Port from LLVM release/13.x. Commits from release_12x (specially CI changes) added on top of the commits. Final commit is mainly changes that required manual merge in order to build the project.
Build procedure is same as release_12x. Test for flang run successfully but some test need to be modified, flang PR also open for review

### CODE REVIEW

File: llvm/lib/CodeGen/AsmPrinter/DwarfDebug.cpp
Comment (pawosm-arm) [inline llvm/lib/CodeGen/AsmPrinter/DwarfDebug.cpp]: Can you give some rationale behind this removal?

File: llvm/lib/CodeGen/AsmPrinter/DwarfDebug.cpp
Comment (pawosm-arm) [inline llvm/lib/CodeGen/AsmPrinter/DwarfDebug.cpp]: Again, can you write some words why this had to go?

File: llvm/lib/CodeGen/AsmPrinter/DwarfDebug.cpp
Comment (abrahamtovarmob) [reply llvm/lib/CodeGen/AsmPrinter/DwarfDebug.cpp]: This change comes from the original code https://github.com/llvm/llvm-project/blob/release/13.x/llvm/lib/CodeGen/AsmPrinter/DwarfDebug.cpp#:~:text=for%20(auto%20%26Entry,List%2C%20BT%2C%20TheCU)%3B
I pushed as-it-is

File: llvm/lib/CodeGen/AsmPrinter/DwarfDebug.cpp
Comment (abrahamtovarmob) [reply llvm/lib/CodeGen/AsmPrinter/DwarfDebug.cpp]: This change comes from the original code https://github.com/llvm/llvm-project/blob/release/13.x/llvm/lib/CodeGen/AsmPrinter/DwarfDebug.cpp#:~:text=List.setTagOffset(*DwarfExpr.TagOffset)%3B

File: llvm/lib/CodeGen/AsmPrinter/DwarfDebug.cpp
Comment (abrahamtovarmob) [reply llvm/lib/CodeGen/AsmPrinter/DwarfDebug.cpp]: Paul commented offline that this change comes from release_12x, I see that the issue was that the base code for release/13.x is different and when I applied the commit for this piece of code the build was breaking, hence I remove the code. What I will do now is to pick up the commit again and do the required changes for release_13x be compatible with it. I will commit-push again once is ready.

File: llvm/lib/CodeGen/AsmPrinter/DwarfDebug.cpp
Comment (abrahamtovarmob) [reply llvm/lib/CodeGen/AsmPrinter/DwarfDebug.cpp]: Paul commented offline that this change comes from release_12x, I see that the issue was that the base code for release/13.x is different and when I applied the commit for this piece of code the build was breaking, hence I remove the code. What I will do now is to pick up the commit again and do the required changes for release_13x be compatible with it. I will commit-push again once is ready.

**Comment by abrahamtovarmob** [comment]
I will temporary close this PR because I'm committing changes to my local branch and I don't want them to reflect here until it's ready. I will reopen until all test are passing

**Comment by pawosm-arm** [comment]
@abrahamtovarmob how much 'ready' this PR is? Seems it is the pre-requisite for everything else in the LLVM13-merge epic.

**Comment by kiranchandramohan** [comment]
@abrahamtovarmob Can you cherry-pick https://github.com/flang-compiler/classic-flang-llvm-project/pull/64 also?

**Comment by bryanpkc** [comment]
We are doing some internal tests of this branch. We will post our feedback when the tests are done.

**Comment by abrahamtovarmob** [comment]
> @abrahamtovarmob Can you cherry-pick #64 also?

Done

**Comment by RichBarton-Arm** [comment]
@kiranchandramohan , @bryanpkc  - are you happy to approve this one? If it looks ok could we merge it then add additional patches on top as separate PRs if needed?

**Review by RichBarton-Arm — APPROVED** [review]
Happy to approve this from the Arm side.

File: clang/lib/CodeGen/BackendUtil.cpp
Comment (bryanpkc) [inline clang/lib/CodeGen/BackendUtil.cpp]: We have replaced `FLANG_LLVM_EXTENSIONS` with `ENABLE_CLASSIC_FLANG` in 56f27ac85775d523757a6826ceb98078972256e7. You probably shouldn't undo that change.

File: clang/lib/Driver/ToolChains/Clang.cpp
Comment (bryanpkc) [inline clang/lib/Driver/ToolChains/Clang.cpp]: We have replaced FLANG_LLVM_EXTENSIONS with ENABLE_CLASSIC_FLANG in 56f27ac. You probably shouldn't undo that change.

File: clang/include/clang/Driver/Types.def
Comment (bryanpkc) [inline clang/include/clang/Driver/Types.def:82]: Lines 81 and 82 used to be disabled if `ENABLE_CLASSIC_FLANG` was set. Now these two lines are processed along with lines 76 though 79. Wouldn't they cause a conflict? A similar problem exists in `clang/lib/Driver/Types.cpp`.

File: clang/lib/Driver/Types.cpp
Comment (bryanpkc) [inline clang/lib/Driver/Types.cpp:273]: These two cases will never match if `ENABLE_CLASSIC_FLANG` is set, right?

File: clang/lib/Driver/Types.cpp
Comment (bryanpkc) [inline clang/lib/Driver/Types.cpp:324]: These cases will never match if `ENABLE_CLASSIC_FLANG` is set, right?

File: llvm/lib/CodeGen/AsmPrinter/DwarfDebug.cpp
Comment (bryanpkc) [inline llvm/lib/CodeGen/AsmPrinter/DwarfDebug.cpp]: Why is this commented out, along with other large sections of code below?

File: llvm/lib/CodeGen/AsmPrinter/DwarfDebug.cpp
Comment (bryanpkc) [inline llvm/lib/CodeGen/AsmPrinter/DwarfDebug.cpp]: Unnecessary indentation here.

File: llvm/lib/CodeGen/AsmPrinter/DwarfDebug.cpp
Comment (bryanpkc) [reply llvm/lib/CodeGen/AsmPrinter/DwarfDebug.cpp]: Consider using `#if 0` instead of `//` to disable large sections of code.

File: llvm/lib/CodeGen/AsmPrinter/DwarfDebug.h
Comment (bryanpkc) [inline llvm/lib/CodeGen/AsmPrinter/DwarfDebug.h:634]: Didn't you comment out the definition of this function in `llvm/lib/CodeGen/AsmPrinter/DwarfDebug.cpp`?

**Comment by bryanpkc** [comment]
You seem to have deleted some test cases:

- `llvm/test/DebugInfo/Generic/fortran-subprogram-at.ll`
- `llvm/test/DebugInfo/Generic/more-subprogram-attr.ll`
- `llvm/test/DebugInfo/X86/DICommonBlock.ll`

Is that intentional?

File: clang/include/clang/Driver/Options.td
Comment (bryanpkc) [inline clang/include/clang/Driver/Options.td:4490]: Question: What is the difference between `f_Group` and `fortran_format_Group` (or `gfortran_Group`)?

**Comment by bryanpkc** [comment]
Nit: A couple of your commit messages have the wording "Port to LLVM release_12x" or "LLVM port for release_12x", which is misleading. I think it's better to say "Port Classic Flang to LLVM 13".

File: clang/lib/CodeGen/BackendUtil.cpp
Comment (abrahamtovarmob) [reply clang/lib/CodeGen/BackendUtil.cpp]: Somehow I left out the commit, thanks for mention it

File: clang/lib/Driver/ToolChains/Clang.cpp
Comment (abrahamtovarmob) [reply clang/lib/Driver/ToolChains/Clang.cpp]: Somehow I left out the commit, thanks for mention it

**Comment by abrahamtovarmob** [comment]
> Nit: A couple of your commit messages have the wording "Port to LLVM release_12x" or "LLVM port for release_12x", which is misleading. I think it's better to say "Port Classic Flang to LLVM 13".

I had changed the name of those commits

File: clang/include/clang/Driver/Options.td
Comment (abrahamtovarmob) [reply clang/include/clang/Driver/Options.td:4490]: f_Group is the original group name that is found in llvm/13.x. Is the Target-independent compilation options
fortran_format_Group is a group for the Fortran source format
gfortran_Group is the gfortran options that we recognize in the driver and pass along when invoking GCC to compile Fortran code.
After initial merge, flang tests fails because it cannot find the symbol, for example, ffixed_form. I don't understand the specifics of each one of it, I based on the code found on release_12x to have an idea how to use the groups names. A more complex merge could be to create a group to encapsulate all three different groups, but as I don't understand specifics of it I cannot tell if that is a good idea and ultimate solution. I vouch for the more simplest solution that was to just change the group name, as it was done between llvm release/12.x and classic release_12x

**Comment by abrahamtovarmob** [comment]
> You seem to have deleted some test cases:
> 
> * `llvm/test/DebugInfo/Generic/fortran-subprogram-at.ll`
> * `llvm/test/DebugInfo/Generic/more-subprogram-attr.ll`
> * `llvm/test/DebugInfo/X86/DICommonBlock.ll`
> 
> Is that intentional?

I didn't got starting commit to cherry pick, thus its omission wasn't intentional

File: clang/include/clang/Driver/Options.td
Comment (RichBarton-Arm) [reply clang/include/clang/Driver/Options.td:4490]: fortran_format_Group is used to help the classic flang driver process the multiple gfortran-style and nvfort/pgfort-style options that control Fortran source format (Mfixed vs -ffixed-form, etc.) I think Abraham's merge here is correct.

In the past, upstream LLVM would have had -ffixed-form in the gfortran_Group to make clang pass it on to gfortran (or ignore it). That functionality broke upstream (I think) and anyway got replaced by new LLVM Flang's support for the options in LLVM 13. LLVM flang places -ffixed-form it in the f_Group (which I think is probably wrong, but that's an upstream bug.) The only implication of that is that the Clang driver (not Flang) will claim all unprocessed f_Group options to prevent an unused argument warning.

File: llvm/lib/CodeGen/AsmPrinter/DwarfDebug.cpp
Comment (abrahamtovarmob) [reply llvm/lib/CodeGen/AsmPrinter/DwarfDebug.cpp]: sorry, fixed

File: llvm/lib/CodeGen/AsmPrinter/DwarfDebug.cpp
Comment (abrahamtovarmob) [reply llvm/lib/CodeGen/AsmPrinter/DwarfDebug.cpp]: Fixed

File: clang/lib/Driver/Types.cpp
Comment (abrahamtovarmob) [reply clang/lib/Driver/Types.cpp:324]: I added the #else exclusion, now these "Case" with TY_Fortran and TY_PP_Fortran wont be match

File: clang/include/clang/Driver/Options.td
Comment (bryanpkc) [reply clang/include/clang/Driver/Options.td:4490]: > The only implication of that is that the Clang driver (not Flang) will claim all unprocessed f_Group options to prevent an unused argument warning.

Thanks @RichBarton-Arm for this explanation. `ToolChains/ClassicFlang.cpp` processes `OPT_fortran_format_Group` but nothing in the code base uses `OPT_gfortran_Group` and `OPT_f_Group`. Could you point me to the Clang driver code that goes through options in `OPT_f_Group` in order to claim them?

Comment (RichBarton-Arm) [reply clang/include/clang/Driver/Options.td:4490]: The bit of code that claims the f_Group options is in Clang's default Driver's ConstructJob function [here](https://github.com/flang-compiler/classic-flang-llvm-project/blob/f4f49d9d0d699f3ac32c1037516c9ab17551991e/clang/lib/Driver/ToolChains/Clang.cpp#L6513)). Seems a shame not to use it to e.g. organise clang's messy --help message to group similar options, but it seems this is all it's used for upstream in clang.

The gfortran support has all completely [bitrotted away](https://reviews.llvm.org/rG6a75496836ea14bcfd2f4b59d35a1cad4ac58cee) upstream in Clang. It will anyway not make much sense when LLVM Flang is  mature. We already agreed that LLVM Flang would have a gfortran-like CLI so we'll want to use these Options.td entries for LLVM Flang options eventually. We might even be able to remove the whole group now. I'll talk to Andrzej about that, but it won't have much bearing on what we're talking about here.

Comment (bryanpkc) [reply clang/include/clang/Driver/Options.td:4490]: Thanks; I didn't realize that `OPT_clang_ignored_f_Group` was related to `OPT_f_Group`. I was trying to understand whether this PR needs to bother changing all the `f_Group` into `gfortran_Group` (for options used by Classic Flang), since it introduces downstream delta that may not really be necessary. If we leave them in `f_Group`, `ToolChains/ClassicFlang.cpp` should claim all of them anyway, and the fact that Clang may try to claim unused copies of such options should be irrelevant.

File: clang/lib/Driver/Types.cpp
Comment (bryanpkc) [reply clang/lib/Driver/Types.cpp:324]: OK thanks.

File: clang/include/clang/Basic/CodeGenOptions.h
Comment (bryanpkc) [inline clang/include/clang/Basic/CodeGenOptions.h]: Please keep the spacing consistent with LLVM 13.

File: clang/include/clang/Frontend/Utils.h
Comment (bryanpkc) [inline clang/include/clang/Frontend/Utils.h]: I suggest keeping this code in sync with [LLVM 13](https://github.com/llvm/llvm-project/blob/release/13.x/clang/lib/Frontend/InitPreprocessor.cpp#L171) (and current main branch), if possible:
```c++
Builder.defineMacro(MacroName, toString(MaxVal, 10, isSigned) + ValSuffix);
```

File: clang/lib/Frontend/InitPreprocessor.cpp
Comment (bryanpkc) [inline clang/lib/Frontend/InitPreprocessor.cpp]: These two `static` versions of `DefineTypeSize` should have been moved to `clang/include/clang/Frontend/Utils.h`.

File: llvm/include/llvm/Analysis/TargetLibraryInfo.h
Comment (bryanpkc) [inline llvm/include/llvm/Analysis/TargetLibraryInfo.h]: Please align the comments with those on surround lines.

File: llvm/lib/CodeGen/AsmPrinter/DwarfDebug.h
Comment (bryanpkc) [reply llvm/lib/CodeGen/AsmPrinter/DwarfDebug.h:634]: Fixed.

File: clang/lib/Driver/ToolChains/Clang.cpp
Comment (bryanpkc) [reply clang/lib/Driver/ToolChains/Clang.cpp]: Fixed.

**Comment by bryanpkc** [comment]
Could you reorder the commits so that they are roughly in the same order as on the release_12x branch? Could you also please squash the `DwarfDebug` changes in 01f56f30430c80f04a86995ef4fad7e4a90f5570 into 885039102d331c9cd8f21c6a8ed2f914b49a785c, and the Types.def and Types.cpp changes in the same commit into 73cafd790b61ca782822be73db790a19a8e26ef4 and then into 769779faf3a1c452941d21308dc76e6467a48caa? I think that will eliminate a lot of the confusing back-and-forth changes, as well as changes that don't seem to belong together in one commit.

**Comment by abrahamtovarmob** [comment]
> Could you reorder the commits so that they are roughly in the same order as on the release_12x branch? Could you also please squash the `DwarfDebug` changes in [01f56f3](https://github.com/flang-compiler/classic-flang-llvm-project/commit/01f56f30430c80f04a86995ef4fad7e4a90f5570) into [8850391](https://github.com/flang-compiler/classic-flang-llvm-project/commit/885039102d331c9cd8f21c6a8ed2f914b49a785c), and the Types.def and Types.cpp changes in the same commit into [73cafd7](https://github.com/flang-compiler/classic-flang-llvm-project/commit/73cafd790b61ca782822be73db790a19a8e26ef4) and then into [769779f](https://github.com/flang-compiler/classic-flang-llvm-project/commit/769779faf3a1c452941d21308dc76e6467a48caa)? I think that will eliminate a lot of the confusing back-and-forth changes, as well as changes that don't seem to belong together in one commit.

Commit order fixed and excessive back-and-forth changes squashed.

**Comment by pawosm-arm** [comment]
Can someone having LLVM12-based (or older) classic flang check one thing for me? I've been trying to examine LLVM-IR files emitted by LLVM13-based flang (with changes from this PR applied locally). What I noticed is that no matter if -g is passed on the command line, the debug info is always present at the end of LLVM-IR file! And it happens at all of the optimization levels. Could it be that something got messed up in the frontend driver and it went unnoticed for some time?!

**Review by shivaramaarao — APPROVED** [review]

**Comment by RichBarton-Arm** [comment]
@bryanpkc  - I have approved from Arm side and Shivram has approved also. If you are happy can you do the push?

**Comment by pawosm-arm** [comment]
I guess flang part must be merged first, otherwise certain test cases will be failing.

**Comment by RichBarton-Arm** [comment]
@pawosm-arm  - only if we enable the flang CI builds with LLVM13 at this point. We can push this then enable CI only once the Flang part has landed.

**Review by bryanpkc — APPROVED** [review]
I have reviewed the changes again and the updated ordering of the commits. LGTM.

**Pull request merged**
